lib/imap/providers/gmail.rb

```diff
@@ -0,0 +1,148 @@
+# frozen_string_literal: true
+
+module Imap
+  module Providers
+    class Gmail < Generic
+      X_GM_LABELS = 'X-GM-LABELS'
+
+      def imap
+        @imap ||= begin
+          imap = super
+          apply_gmail_patch(imap)
+          imap
+        end
+      end
+
+      def emails(mailbox_name, uids, fields)
+        fields[fields.index('LABELS')] = X_GM_LABELS
+
+        emails = super(mailbox_name, uids, fields)
+
+        emails.each do |email|
+          email['LABELS'] = Array(email['LABELS'])
+
+          if email[X_GM_LABELS]
+            email['LABELS'] << Array(email.delete(X_GM_LABELS))
+            email['LABELS'].flatten!
+          end
+
+          email['LABELS'] << '\\Inbox' if mailbox_name == 'INBOX'
+
+          email['LABELS'].uniq!
+        end
+
+        emails
+      end
+
+      def store(uid, attribute, old_set, new_set)
+        attribute = X_GM_LABELS if attribute == 'LABELS'
+        super(uid, attribute, old_set, new_set)
+      end
+
+      def to_tag(label)
+        # Label `\\Starred` is Gmail equivalent of :Flagged (both present)
+        return 'starred' if label == :Flagged
+        return if label == '[Gmail]/All Mail'
+
+        label = label.to_s.gsub('[Gmail]/', '')
+        super(label)
+      end
+
+      def tag_to_flag(tag)
+        return :Flagged if tag == 'starred'
+
+        super(tag)
+      end
+
+      def tag_to_label(tag)
+        return '\\Important' if tag == 'important'
+        return '\\Starred' if tag == 'starred'
+
+        super(tag)
+      end
+
+      private
+
+      def apply_gmail_patch(imap)
+        class << imap.instance_variable_get('@parser')
```

Comment:
The logic is a little hard to follow so can you clarify something for me here? Does this patch the `Imap` class, so that every instance of the imap will now be patched for gmail? Or does it only patch the specific instance?

Reply:
Yes, it opens the metaclass (/ eigenclass / singleton class) of the parser instance to add support for Gmail's IMAP extensions (`X-GM-*` attributes).